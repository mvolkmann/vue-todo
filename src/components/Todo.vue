<template>
  <li>
    <input type="checkbox" :checked="todo.done" @change="onToggleDone">
    <span :class="doneClass">{{todo.text}}</span>
    <button @click="onDeleteTodo">Delete</button>
  </li>
</template>

<script>
/* eslint-disable no-console */
export default {
  name: 'Todo',
  props: {
    done: Boolean,
    onDeleteTodo: {
      type: Function,
      required: true
    },
    onToggleDone: {
      type: Function,
      required: true
    },
    todo: {
      type: Object,
      required: true,
      validator(obj) {
        return obj.text && obj.done !== undefined;
      }
    }
  },
  computed: {
    doneClass() {
      return 'done-' + this.todo.done;
    }
  }
};
</script>

<style lang="scss" scoped>
$done-color: gray;

button {
  margin-left: 10px;
}

li {
  margin-top: 5px;

  .done-true {
    color: $done-color;
    text-decoration: line-through;
  }
}
</style>
