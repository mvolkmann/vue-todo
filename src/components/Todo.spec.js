import {mount, shallowMount} from '@vue/test-utils';
import Todo from '../../src/components/Todo';

describe('Todo', () => {
  const text = 'buy milk';
  const todo = {text, done: false};

  test('should match snapshot', () => {
    const wrapper = mount(Todo, {
      propsData: {onDeleteTodo() {}, onToggleDone() {}, todo}
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  test('should render', () => {
    const wrapper = shallowMount(Todo, {
      propsData: {onDeleteTodo() {}, onToggleDone() {}, todo}
    });
    const checkbox = wrapper.find('input[type="checkbox"]');
    expect(checkbox).not.toBeNull();
    const html = wrapper.html();
    expect(html).toContain(text); // the todo text
    expect(html).toContain('Delete'); // the button
  });

  test('should handle Delete button', () => {
    let called = false;
    const wrapper = shallowMount(Todo, {
      propsData: {
        onDeleteTodo() {
          called = true;
        },
        onToggleDone() {},
        todo
      }
    });
    const deleteBtn = wrapper.find('button');
    deleteBtn.trigger('click');
    expect(called).toBe(true);
  });

  test('should toggle done', () => {
    let called = false;
    const wrapper = shallowMount(Todo, {
      propsData: {
        onDeleteTodo() {},
        onToggleDone() {
          called = true;
        },
        todo
      }
    });
    const checkbox = wrapper.find('input[type="checkbox"]');
    checkbox.trigger('click');
    expect(called).toBe(true);
  });
});
