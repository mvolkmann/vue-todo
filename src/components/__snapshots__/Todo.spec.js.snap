// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo should match snapshot 1`] = `
<li>
  <input
    type="checkbox"
  />
   
  <span
    class="done-false"
  >
    buy milk
  </span>
   
  <button
    class="delete-btn"
  >
    Delete
  </button>
</li>
`;
